This file contains the code for cleaning the rounds2 file of the project.

## Code Style
    - Case: 
        - snake_case for objects
        - camelCase for functions and classes
    - Double quotes first, then single quotes

## Libraries used
    - Pandas
    - Numpy

## Obejctives of Analysis
Identify the most heavily invested main sectors in each of the three countries (for funding type FT and investments range of 5-15 M USD).

Business objective: Identify the best: a. Sectors; b. Countries; c. Investment rounds for Spark Funds.

[This means that we need to focus on just a few variables]

## Metric
Mean amount of money invested in a particular country. 

## The Workflow
The workflow for this analysis is rather simple. Focus on answering the questions asked in the checkpoints. Following this flow, the code in this .ipynb is organized according to the checkpoints. There will be a clear heading indicating the starting and ending of each checkpoint and question.

In [1]:
# importing dependencies
# numpy
import numpy as np # version: 1.15.0

# pandas
import pandas as pd # version: 0.23.4

/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


# Checkpoint 1: Data Cleaning
There are five tasks in this checkpoint:
    - Number of unique companies in rounds2.csv
	- Number of unique companies in companies.tsv
	- Key column from the companies dataset that can be used to merge it with rounds data
	- Organizations in companies that are missing in rounds2.
    - Merge the two datasets.

## Importing the data
The first step of the analysis is to import the two main datasets that we will be needing for the analysis: companies and rounds. 

In [2]:
# import companies.csv as companies
companies = pd.read_csv("../../Data/companies.tsv", sep = "\t", encoding = "ISO-8859-1") 

# import rounds2.csv as rounds
rounds = pd.read_csv("../../Data/rounds2.csv", sep = ",", encoding = "ISO-8859-1")
# ISO for lack of charset in UTF-8

In [3]:
#information of the companies dataset
print(companies.info()); print("shape of dataset: ", companies.shape); print("variable dtypes:\n", companies.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB
None
shape of dataset:  (66368, 10)
variable dtypes:
 permalink        object
name             object
homepage_url     object
category_list    object
status           object
country_code     object
state_code       object
region           object
city             object
founded_at       object
dtype: object


In [4]:
# information about the rounds dataset
print(rounds.info()); print("shape of dataset: ", rounds.shape); print("variable dtypes:\n", rounds.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB
None
shape of dataset:  (114949, 6)
variable dtypes:
 company_permalink           object
funding_round_permalink     object
funding_round_type          object
funding_round_code          object
funded_at                   object
raised_amount_usd          float64
dtype: object


Since we will be focusing mostly on four variables only, let's remove all the extraneous variables from both the datasets. 
We'll remove from the companies dataset the following variables:
    - state_code
    - region
    - city
    - homepage_url
    - founded_at
    - name

In [5]:
# removing unnecessary columns from companies
companies.drop(["state_code", "region", "city", "homepage_url", "founded_at", "name"], axis = 1, inplace = True)

We'll remove the following from rounds:
    - funded_at
    - funding_round_code
    - funding_round_permalink

In [6]:
# removing unnecessary columns from rounds
rounds.drop(["funded_at", "funding_round_code", "funding_round_permalink"], axis = 1, inplace = True)

## Checkpoint 1 Q1: Number of unique companies in rounds
To do this, we'll use the company_permalink column. However, instead of doing this directly, we'll first convert the company_permalink to lowercase and then determine the number of unique records.

In [7]:
# converting company_permalink to lower case and getting number of unique records.
rounds.company_permalink.str.lower().nunique()

66370

There seem to be 66370 unique companies in the dataset. This means that there are companies that had more than one round of funding. [Import Observation]

## Checpoint 1 Q2: Number of unique companies in companies
This time, we'll use the permalink, which is supposed to be the UID of a company. As with rounds.company_permalink, we'll first convert to lower case and then proceed to count the number of unique records.

In [8]:
companies.permalink.str.lower().nunique()

66368

There seems to be a discrepancy between the number of unique records in companies and rounds. Does this mean that there are at least 2 companies in rounds that are not present in companies?

## Checkpoint 1 Q3: Key column to merge companies and rounds
This is pretty easy. From the data dictionary we know the companies.permalink and rounds.company_permalink are UID's of each company in the dataset. So, we'll use companies.permalink as the key to merge with rounds.

## Checkpoint 1 Q4: Mismatches between rounds and companies
Ok. Now, we're required to find out if there are any records that are unique to rounds only. That is these organizations are not present in companies but are present in rounds.

We can do this by merging on companies.permalink and rounds.company_permalink. But, we'll take a slightly different approach here. 

First off, we'll create two new columns in rounds and companies called company_name and name resp. Then, we'll merge based on those columns and check for missing values. If there are missing values, then there are companies which are unique to rounds only.

# Update
So, I've found out that using a case-unified form of the permalinks produces the same result as using the names. Thus, to avoid creating unnecessary variables and excess storage consumption, I'll hold off on creating those extra columns and use the lower-case permalinks themselves.

In [9]:
# converting rounds.company_permalink and companies.permalink to lower case
companies["permalink"] = companies.permalink.str.lower()

rounds["company_permalink"] = rounds.company_permalink.str.lower()

In [10]:
# checking if there are any unique records.
print(rounds.company_permalink[~rounds.company_permalink.isin(companies.permalink)].dropna());

29597                               /organization/e-cãŠbica
31863          /organization/energystone-games-灵石游戏
45176                  /organization/huizuche-com-惠ç§ÿ车
58473                /organization/magnet-tech-磁石科技
101036    /organization/tipcat-interactive-沙èˆÿ信息ç...
109969               /organization/weiche-tech-喂车科技
113839                   /organization/zengame-禅游科技
Name: company_permalink, dtype: object


(Look at this stackoverflow answer: https://stackoverflow.com/questions/28901683/pandas-get-rows-which-are-not-in-other-dataframe for the full explanation of the code used above.)

There seem to be 7 companies that are in rounds but not in companies. So, the answer to the fourth question is yes. There are organizations that are present in rounds but not in companies.

## Checkpoint 1 Q5: Merge the two dataframes
This is the basis of all our analysis. Merging the two DataFrames will give us a single data frame which contains all the data needed. After this step, we can finally start treating the missing values.

So, after some thought it was decided unanimously that the best approach would be to drop all missing values from the raised_amount_usd column. Therefore, that's what I'll do.

In [11]:
# dropping all missing values from the rounds DataFrame.
rounds.dropna(inplace = True)

In [12]:
rounds.shape # new shape of rounds: (94959, 3)

(94959, 3)

Dropping those missing values leaves us with 95K observations. So, we lost about 20K observations. Not a big problem though. The next step is to combine the two datasets.

In [13]:
master = pd.merge(rounds, companies, how = "inner", left_on = "company_permalink", 
                  right_on = "permalink")

So, performing this merge resulted in a DataFrame with 94958 rows and 7 columns.

In [14]:
# checking for missing values
master.isnull().mean().sort_values(ascending = False)

country_code          0.061606
category_list         0.010994
status                0.000000
permalink             0.000000
raised_amount_usd     0.000000
funding_round_type    0.000000
company_permalink     0.000000
dtype: float64

There are two variables with missing values: country_code (an important variable) has 6% values missing and category_list has 1% values missing. 

The next thing to do is to check for duplicates. If there are duplicates, just drop them.

In [15]:
# checking for duplicates in the master DataFrame
master.duplicated().sum() # there seem to be 1311 duplicated values.

1311

In [16]:
master.drop_duplicates(keep = "first", inplace = True)

So, dropping duplicates resulted in the loss of about 2K rows. We're now left with about 93K rows. This is about 81% of the initial observations. I don't think that deleting any more values is a good idea. This would result in the loss of more data, bringing the total down even more. 

Still, if we can retain more than 75% of the rows, we will have pretty reliable results.

# Treating the remaining missing values
The remaining missing values are both present in Series of dtype object. This means that we can do something cool. We can treat the missing value itself as a new category. This is sweet. This means that we don't need to lose any more data. So, let's go ahead and do that. What we'll do is replace the NaN's with a new sentinel value that tells us if we know the country_code of an organization or not.

In [17]:
# replacing missing values in country_code with new sentinel value
master["country_code"].fillna("unknown", inplace = True)

In [18]:
# on running the following code, we see that the missing value percentage has gone down
master.isnull().mean().sort_values(ascending = False)

category_list         0.011106
country_code          0.000000
status                0.000000
permalink             0.000000
raised_amount_usd     0.000000
funding_round_type    0.000000
company_permalink     0.000000
dtype: float64

Now, we have about 1% of the total values missing 1% of it's data. 1% of 93647 is about 940. We can go ahead and delete these rows. We'll still be left with 80.5% of the total observations.

In [19]:
master.dropna(inplace = True) # Now that we have this DataFrame, we can finally start with the analysis.

# Checkpoint 2: Funding Type Analysis

In [20]:
# checking the information about the master table after performing all the operations
print(master.info()); print(master.shape)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 92607 entries, 0 to 94957
Data columns (total 7 columns):
company_permalink     92607 non-null object
funding_round_type    92607 non-null object
raised_amount_usd     92607 non-null float64
permalink             92607 non-null object
category_list         92607 non-null object
status                92607 non-null object
country_code          92607 non-null object
dtypes: float64(1), object(6)
memory usage: 5.7+ MB
None
(92607, 7)


# Checkpoint 3: Country Analysis

# Checkpoint 4: Sector Analysis 1
So, let's start off with the sector analysis. Here's what needs to be done in this part of the analysis.
    - Extract the primary sector of each category list from the category_list column.
    - Use the mapping file 'mapping.csv' to map each primary sector to one of the eight main sectors (Note that ‘Others’ is also considered one of the main sectors)

The primary sector is the string that appears before the first pipe "|" in the category_list variable. So, let's get on with it.

## Extracting the primary sector from category_list
First off, let's start by getting a good look at the dataframe. After that, we'll create a new variable called primary_sector to store the primary sector of each organization in the dataframe.

In [21]:
print(master.info()); print(master.shape) # cool

<class 'pandas.core.frame.DataFrame'>
Int64Index: 92607 entries, 0 to 94957
Data columns (total 7 columns):
company_permalink     92607 non-null object
funding_round_type    92607 non-null object
raised_amount_usd     92607 non-null float64
permalink             92607 non-null object
category_list         92607 non-null object
status                92607 non-null object
country_code          92607 non-null object
dtypes: float64(1), object(6)
memory usage: 5.7+ MB
None
(92607, 7)


In [22]:
# creating the primary sector
master["primary_sector"] = master.category_list.str.split("|").str.get(0)

Now that we have the primary sector, the next to-do item on our list is to map each of those primary sectors to a main sector. For that, we need the mapping data. We'll load that up and proceed to map the primary sectors to a main sector.

In [23]:
# loading mapping.csv as mapping
mapping = pd.read_csv("../../Data/mapping.csv", sep = ",", encoding = "ISO-8859-1")

In [24]:
# some basic information about mapping
print(mapping.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 688 entries, 0 to 687
Data columns (total 10 columns):
category_list                              687 non-null object
Automotive & Sports                        688 non-null int64
Blanks                                     688 non-null int64
Cleantech / Semiconductors                 688 non-null int64
Entertainment                              688 non-null int64
Health                                     688 non-null int64
Manufacturing                              688 non-null int64
News, Search and Messaging                 688 non-null int64
Others                                     688 non-null int64
Social, Finance, Analytics, Advertising    688 non-null int64
dtypes: int64(9), object(1)
memory usage: 53.8+ KB
None


In [25]:
mapping.head()

category_list  Automotive & Sports  Blanks  Cleantech / Semiconductors  \
0            NaN                    0       1                           0   
1             3D                    0       0                           0   
2    3D Printing                    0       0                           0   
3  3D Technology                    0       0                           0   
4     Accounting                    0       0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1

OK. So, this mapping file has turned out to be a one-to-one sparse matrix that maps each primary sector to one of the eight main sectors. What I have to do now is get a main sector for each primary sector. I'm not aware of any native pandas methods that enable me to do this. My first impulse is to write a function which will allow me to do just this. Then there's relational algebra which can produce results really quickly. I'll try out my function first.

As it turns out, this operation is what Hadley Wickham calls gathering the columns. (read more about it here: https://r4ds.had.co.nz/tidy-data.html).

The main task here is this: convert the wide (and sparse) representation of the mapping into a long representation. This means bringing in all the colums under one roof. It's way easier to demonstrate than to explain. 

Luckily, pandas does provide a function to do this: pd.melt() (just like reshape2::melt() from R). Check the docs to understand how awesome this function is. (For a tutorial on pd.melt() go here: https://www.ibm.com/developerworks/community/blogs/jfp/entry/Tidy_Data_In_Python?lang=en)

In [26]:
# dropping the Blanks column since this just serves as a flag to identify NaN's
mapping.drop("Blanks", axis = 1, inplace = True)

# dropping the single NaN at the head of the dataset
mapping.dropna(axis = 0, inplace = True)

In [27]:
# gathering all the columns under one roof
mapping_long = pd.melt(mapping, id_vars = ["category_list"], var_name = "main_sector", value_name = "yes_no")

# tidying mapping_long to produce the final version of the mapping dataset
mapping_tidy = mapping_long.loc[mapping_long.yes_no == 1, ["category_list", "main_sector"]]

To finish up, let's take one look at the mapping_tidy dataset just to make sure everything is alright

In [28]:
print(mapping_tidy.info());
print(mapping_tidy.isnull().mean()) # no null values. We can proceed to merge the two datasets!

<class 'pandas.core.frame.DataFrame'>
Int64Index: 687 entries, 7 to 5471
Data columns (total 2 columns):
category_list    687 non-null object
main_sector      687 non-null object
dtypes: object(2)
memory usage: 16.1+ KB
None
category_list    0.0
main_sector      0.0
dtype: float64


Now, all that's left out is to merge the two datasets and proceed to perform a series of checks. Here are the final steps to finish up checkpoint 4.
1. Merge master and mapping_tidy
2. Check for: a. Missing values; b. Duplicates
3. Treat missing values and drop duplicates`

# Issue with merging
Now, while merging the two datasets, master and mapped_tidy, one of two types of joins can be used. Either the inner or outer. There are the left and right joins, but we'll leave them out of the picture for now. 

If the inner join is used during the merge, there is a loss data. There are some sub-sectors that are present in master.primary_sector but aren't present in mapped.category_list. 

If the outer join is used during the merge, NULL values are inserted into the dataset. Again, dropping this is just equivalent to using the inner join. One other way to treat them is to manually add them to one of the 8 main sectors. 

What do you guys think?

In [29]:
# primary sectors present in master.primary_sector but not present in mapping_tidy.category_list
master.primary_sector[~master.primary_sector.isin(mapping_tidy.category_list)].dropna().unique()

array(['Analytics', 'Finance', 'Financial Services', 'Nanotechnology',
       'Finance Technology', 'Business Analytics', 'Contact Management',
       'Career Management', 'Big Data Analytics', 'Self Development',
       'Cloud Management', 'Digital Signage', 'Cause Marketing',
       'Personalization', 'Innovation Management', 'Risk Management',
       'Document Management', 'Journalism', 'Energy Management',
       'Waste Management', 'Fleet Management', 'Alternative Medicine',
       'Real Estate Investors', 'Investment Management',
       'English-Speaking', 'Intellectual Asset Management',
       'Educational Games', 'Identity Management', 'Lead Management',
       'Property Management', 'IT Management', 'Event Management',
       'Navigation', 'Toys', 'Professional Services', 'Generation Y-Z',
       'Cannabis', 'Enterprise Hardware', 'Social Media Advertising',
       'Personal Finance', 'Darknet', 'Knowledge Management',
       'China Internet', 'Medical Professionals',
       

From the result of the code above, it's clear that there are 89 sub-sectors that are not included in the mapping dataset. So, if we're manually imputing the values, it would mean the addition of utmost, 89 lines of code. This would preserve the data, but decrease the readability of the file quite a bit. Since we're graded on code readability quite a bit, what do you think we should do?

To deal with this merging issue, here's what I've decided. By observation, I've found out that the following primary_sectors consitute about 90% of the missing values in the dataset. So, I've decided to place these primary_sectors into one of the 8 main_sectors. Thereby, significantly reducing the data loss that will occur if we drop the missing data. 

These are primary sectors:
Social, Analytics, Finance, Advertising:
    - Analytics 
    - Finance 
    - Financial Services 
    - Finance Technology 
    - Business Analytics 
    - Big Data Analytics 
    - Investment Management 
    - Social Media Advertising 
    - Personal Finance 
    - Predictive Analytics 
    - Financial Exchanges
    - Mobile Analytics 
    - Social Media Management
    - Promotional
Cleantech / Semiconductors:
    - Waste Management
    - Natural Gas Uses
    - Biotechnology and Semiconductor
    - Green Tech
    - Energy Management
    - Natural Resources
Health:
    - Alternative Medicine
    - Cannabis
    - Medical Professionals
    - Personal Health
    - Mobile Emergency&Health
(I've segregated each of these primary sectors into one of the eight main sectors based on what I think is appropriate. Please go through the list and let me know if there are any changes.)

In [30]:
# creating lists of primary_sectors that fall under a main sector
# Social, Finance, Analytics, Advertising
social_analytics = ["Analytics", "Finance", "Financial Services", "Finance Technology", "Business Analytics", 
"Big Data Analytics", "Investment Management", "Social Media Advertising", "Personal Finance", 
"Predictive Analytics", "Financial Exchanges", "Mobile Analytics", "Social Media Management", "Promotional"]

# Cleantech / Semiconductors
cleantech_semiconductors = ["Waste Management", "Natural Gas Uses", "Natural Resources", 
"Biotechnology and Semiconductor", "GreenTech", "Energy Management"]

# Health
health = ["Alternative Medicine", "Cannabis", "Meidcal Professionals",
"Personal Health", "Mobile Emergency&Health"]

Now that the bins are ready, we'll perform the following steps in order:
1. Merge the mapping_tidy dataset with the master dataset using an left join
2. Impute the missing main_sector values in the merged master dataset
3. Replace the other missing values in main_sector with the "Others" flag.

### Why I think deleting missing values is a bad idea at this stage?
The answers for questions in Checkpoints 2 and 3 were found out with a dataset containing 92607 rows. Because of that, deleting any missing values, which might lead to the deletion of rows might bias the analysis and produce erroneous results. Therefore, I feel that it is better to flag the remaining 1000 rows as "Others" than to remove those rows.

In [31]:
# merging the master and mapping_tidy
master = pd.merge(master, mapping_tidy, how = "left", left_on = "primary_sector", right_on = "category_list")

In [32]:
# removing the category_list_y variable and renaming category_list_x as category_list
master.drop("category_list_y", axis = 1, inplace = True)

master.rename(index = str, columns = {"category_list_x": "category_list"}, inplace = True)

In [33]:
# filling up the main_sector with appropriate values.
# social, analytics, finance, advertising
master.loc[master.primary_sector.isin(social_analytics), "main_sector"] = "Social, Finance, Analytics, Advertising"

# cleantech / semiconductors
master.loc[master.primary_sector.isin(cleantech_semiconductors), "main_sector"] = "Cleantech / Semiconductors"

# health
master.loc[master.primary_sector.isin(health), "main_sector"] = "Health"

# filling up the remaining missing values with others
master.loc[master.main_sector.isnull(), "main_sector"] = "Others"

Now, the dataset can be used to for the analysis.

In [34]:
# sample code. Not for the final file. 
# converting raised_amount_usd to millions (divide 1000000)
# master["raised_amount_usd"] = master.raised_amount_usd / 1000000
# master.groupby("main_sector").raised_amount_usd.mean().sort_values(ascending = False)
# master.groupby("funding_round_type").raised_amount_usd.mean().sort_values(ascending = False)

# Checkpoint 5: Sector Analysis 2